src/linear_algebra/matrix/polynomial.lean

```diff
@@ -0,0 +1,109 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import linear_algebra.matrix.determinant
+import data.polynomial.eval
+import data.polynomial.monic
+import algebra.polynomial.big_operators
+
+/-!
+# Matrices of polynomials and polynomials of matrices
+
+In this file, we prove results about matrices over a polynomial ring.
+In particular, we give results about the polynomial given by
+`det (t * I + A)`.
+
+## References
+
+  * "The trace Cayley-Hamilton theorem" by Darij Grinberg, Section 5.3
+
+## Tags
+
+matrix determinant, polynomial
+-/
+
+open_locale matrix big_operators
+
+variables {n α : Type*} [decidable_eq n] [fintype n] [comm_ring α]
+
+open polynomial matrix equiv.perm
+
+namespace polynomial
+
+lemma nat_degree_det_X_add_C_le (A B : matrix n n α) :
+  nat_degree (det ((X : polynomial α) • A.map C + B.map C)) ≤ fintype.card n :=
+begin
+  rw det_apply,
+  refine (nat_degree_sum_le _ _).trans _,
+  refine (multiset.max_le_of_forall_le _ _ _),
+  { simp only [forall_apply_eq_imp_iff', true_and, function.comp_app, multiset.map_map,
+               multiset.mem_map, exists_imp_distrib, finset.mem_univ_val],
+    intros g,
+    suffices : (∏ (i : n), (X • A.map C + B.map C) (g i) i).nat_degree ≤
+      fintype.card n,
+    { cases int.units_eq_one_or (sign g) with sg sg,
+      { rw [sg, one_smul],
+        exact this },
+      { rw [sg, units.neg_smul, one_smul, nat_degree_neg],
+        exact this } },
+    refine (nat_degree_multiset_prod_le _).trans _,
+    refine (multiset.sum_le_of_forall_le _ 1 _).trans _,
+    { simp only [forall_apply_eq_imp_iff', true_and, algebra.id.smul_eq_mul, function.comp_app,
+                 ring_hom.map_matrix_apply, multiset.map_map, map_apply, multiset.mem_map,
+                 dmatrix.add_apply, pi.smul_apply, exists_imp_distrib, finset.mem_univ_val],
+      intro,
+      refine (nat_degree_add_le _ _).trans _,
+      simpa using (nat_degree_mul_C_le _ _).trans nat_degree_X_le },
+    { simpa [fintype.card] } }
```

Comment:
Wouldn't this look nicer / be clearer as a `calc` block? I'll cook one up soon if you haven't already by then.

Reply:
Here is an attempt. I couldn't do calc on the top few lines because they introduce an implication step that `simp only` then converts back into an inequality. The same happens below, which results in the two calc blocks.